### ISSUE TITLE
Gradient w.r.t. parameters or modules

### ISSUE THREAD

**Issue by jackshi0912** [pytorch/pytorch#74832]
### 🚀 The feature, motivation and pitch

Often you have multiple models and multiple losses in a system. GAN for example typically has a generator and a discriminator and uses a discriminator loss and a generator loss. 
In these cases, it's often the case that you want to compute gradient of a specific loss w.r.t. a specific model. Again in GAN, you want to compute discriminator loss w.r.t. the discriminator and generator loss w.r.t. the generator.

Right now, this is not possible. Let's say you have generator G, discriminator D, and noise z vector z.
Example from official pytorch DCGAN Tutorial: https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html#training
```
fake = G(z)
discriminator_signal = D(fake.detach())
discriminator_loss = loss1(discriminator_signal)
discriminator_loss.backward
// Optimise discriminator

discriminator_signal_for_generator = D(fake)
generator_loss = loss2(discriminator_signal_for_generator)
generator_loss.backward
// Optimise generator
```

In this case you have to do a discriminator forward pass twice on the same data which is completely unnecessary. This is because there are 2 different losses targeting at 2 different modules. Tensor::backward cannot specify the modules for which gradient should be computed. If omit the `detach` and use `zero_grad` to avoid 2 passes of the discriminator, then we would compute generator grad for `discriminator_loss` for no reason.

Instead, we could have the following
```
discriminator_signal = D(G(z))
discriminator_loss = loss1(discrminator_signal)
generator_loss = loss2(discriminator_signal)
discriminator_loss.backward(D)  # or discriminator_loss.backward(D.parameters)
generator_loss.backward(G)  # or generator_loss.backward(G.parameters)
```

This is the simplest example. In bigger systems such as RL agents with 5 or 6 losses (including adversarial components like MGAIL), we end up doing many unnecessary passes.

This feature can be implemented as simple as detaching everything else while computing gradient. It would be great if this can also be implemented in libtorch.

Please let me know what you think!

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Hi,

Thanks for the details.
Isn't the `inputs` argument to `.backward()` doing exactly what you want? (doc here: https://pytorch.org/docs/stable/generated/torch.autograd.backward.html?highlight=backward#torch.autograd.backward)

**Comment by jackshi0912** [comment]
Yes thank you, I completely missed it!
